### ISSUE TITLE
Continuous legend

### ISSUE THREAD

**Issue by grantmcdermott** [grantmcdermott/tinyplot#84]
For numeric variables with many categories.

See suggestions here: https://stackoverflow.com/questions/13355176/gradient-legend-in-base

**Comment by grantmcdermott** [comment]
I was playing around with some ideas over the weekend and I think it's possible to a hack [this approach](https://stackoverflow.com/a/58189259/4115816) in way that (a) integrates with the existing `plot2::draw_legend()` logic and (b) looks pretty reasonable.

Proof of concept:

``` r
library(plot2)

# some "global" params and palette options for this MWE

ncolors = 101L
nlabs = 5L
collab = "Sepal.Length"

pal = hcl.colors(ncolors, "inferno")
palramp = colorRampPalette(pal)

colvar = iris[[collab]]

# colors for the actual plot points

cols = palramp(length(colvar))
## NB: we reorder now so we can assign colors as part of the points()
## call later on. (Saves having to split by a continuous variable.)
colord = findInterval(colvar, sort(colvar))
cols = cols[colord]

# legend color ramp

## subset to the pretty labels for the legend
pcolvar = sort(pretty(colvar, n = nlabs), decreasing = TRUE)
# pcolvar = pcolvar[pcolvar >= min(colvar) & pcolvar <= max(pcolvar)]
colvar2 = unique(c(colvar, pcolvar))
lcolord = findInterval(pcolvar, sort(colvar2))
cols2 = palramp(length(colvar2))
lcols = cols2[lcolord]
lcols_all = colorRampPalette(lcols)(ncolors)

# legend labels

lgd = rep(NA, times = ncolors)
## match the legend labels to their position in the colorbar
pidx = na.omit(match(lcols, lcols_all))
if (!is.null(attr(pidx, "na.action"))) {
    lgd[as.integer(pidx)] = pcolvar[-attr(pidx, "na.action")]
} else {
    lgd[pidx] = pcolvar
}
## Add "padding" on either side, otherwise the y.intersp adjustment
## (see below) will cause the labels to look funny
lgd = c(NA, lgd, NA)
lcols_all = c(NA, lcols_all, NA)

# draw the legend using our bespoke colour ramp and labels

## first adjust vertical legend spacing (to compress as if there were only 'nlabs'
## discrete colour boxes to draw)
intersp_adj = rep(1/ncolors*length(pcolvar), times = ncolors)
intersp_adj[1] = 1

## now draw the legend
draw_legend(
    legend = "right!",
    lgnd_labs = lgd,
    legend.args = list(
        title = collab,
        bty = "o",
        y.intersp = intersp_adj,
        pt.cex = 3.5
    ),
    type = "p",
    pch = 22,
    col = NA,
    bg = lcols_all
)

# Manually add the points and other plot region elements
# (mostly to give an idea of what the end plot will look like)

plot.window(xlim = range(iris$Petal.Length), ylim = range(iris$Sepal.Length))
axis(1); axis(2); grid()
title(xlab = "Petal.Length", ylab = "Sepal.Length", main = "Continuous legend")
points(
    x = iris$Petal.Length, y = iris$Sepal.Length,
    pch = 19,
    col = cols
)
```

![](https://i.imgur.com/7KgiJ1n.png)

<sup>Created on 2024-01-29 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by grantmcdermott** [comment]
Better version that uses a rescaling function to get the colour mapping right. (Helps to avoid finicky label matching code too.)

``` r
library(plot2)

# rescaling function
rescale_func = function (x, from = NULL, to = NULL) {
    if (is.null(from)) from = range(x)
    if (is.null(to)) to = c(1, 100)
    (x - from[1])/diff(from) * diff(to) + to[1]
}

# data and input columns

dataset = iris
xlab = "Petal.Length"
ylab = "Sepal.Length"
collab = "Sepal.Length"

xvar = dataset[[xlab]]
yvar = dataset[[ylab]]
colvar = dataset[[collab]]

# color palette choices and funcs

ncolors = 100L
nlabs = 5L
pal = hcl.colors(ncolors, "inferno", alpha = 1)
palramp = colorRampPalette(pal, alpha = TRUE)

# generate 'ncolors' (here: 100) distinct color categories for the plot
cols = palramp(ncolors)

# legend color ramp

## Identify the pretty break points for our labels
ucolvar = unique(colvar)
colvar_range = range(ucolvar)
pcolvar = pretty(colvar_range, n = nlabs)
pcolvar = pcolvar[pcolvar >= colvar_range[1] & pcolvar <= colvar_range[2]]
# if (length(pcolvar)>nlabs) pcolvar = pcolvar[-1] # optional
# optional thinning
if (length(ucolvar)==2 && all(ucolvar %in% pcolvar)) {
    pcolvar = ucolvar
} else if (length(pcolvar)>nlabs) {
    pcolvar = pcolvar[seq_along(pcolvar) %% 2 == 0]
}

## Find the (approximate) location of our pretty labels
pidx = rescale_func(c(colvar_range, pcolvar), to = c(1, ncolors))[-c(1:2)]
pidx = round(pidx)
lgd_labs = rep(NA, times = length(cols))
lgd_labs[pidx] = pcolvar
# We have to reverse the order since the legends are in decreasing sequence
lgd_labs = rev(lgd_labs)
lgd_cols = rev(cols)

# Add "padding" on either side, otherwise the y.intersp adjustment
# below will cause the labels to look funny
lgd_labs = c(NA, lgd_labs, NA)
lgd_cols = c(NA, lgd_cols, NA)

## draw the legend using our bespoke color ramp and pretty labels

intersp_adj = rep(1 / ncolors * nlabs, times = length(lgd_cols))
intersp_adj[1] = 1
intersp_adj[length(intersp_adj)] = 1
draw_legend(
    legend = "right!",
    lgnd_labs = lgd_labs,
    legend.args = list(
        title = collab,
        # bty = "o",
        y.intersp = intersp_adj,
        adj = c(0,0),
        pt.cex = 3.5
    ),
    type = "p",
    pch = 22,
    col = NA,
    bg = lgd_cols
)

# Manually add the points and other plot region elements
# (mostly to show what the final plot would look like)

plot.window(xlim = range(xvar), ylim = range(yvar))
axis(1); axis(2); grid()
title(xlab = xlab, ylab = ylab, main = "Continuous legend")
points(
    x = xvar, y = yvar,
    pch = 19,
    col = cols[round(rescale_func(colvar))]
)
```

![](https://i.imgur.com/NcepK5z.png)

<sup>Created on 2024-02-01 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by grantmcdermott** [comment]
Note to self: This `intersp` trick probably won't work for horizontal legends (incl. "top!" and "bottom!"). We can just exclude those for now.

**Comment by grantmcdermott** [comment]
@zeileis As our [resident colour expert](https://journal.r-project.org/articles/RJ-2023-071/): Do you have a recommendation for the default continuous palette? Right now, it's "viridis" but I'm not the _biggest_ fan TBH. Do you perhaps think "cividis", or one of the single palette HCL colors like "blues3" would be a better choice? Something else?

Background: I have the basic code working for native continuous legends in tinyplot now. So it seems a good time to decide the default palette. Happy to share some screenshots if that's helpful.

**Comment by zeileis** [comment]
It depends what you want to use the palette for.

- If it is for shading areas (e.g., in a heatmap), then the viridis palettes typically have too much chroma. Then it is better to have a palette like "YlGnBu" that only has high chroma for the dark colors and low chroma for the light colors. See
[bottom left panel in Figure 5](https://journal.R-project.org/articles/RJ-2023-071/#fig-dorian) from the paper you mention above. The comparison in [Figure 6](https://journal.r-project.org/articles/RJ-2023-071/#fig:ylgnbu-viridis) also brings out the contrast between YlGnBu and Viridis quite clearly.
- However, if you use the continuous palette for coloring points or lines (rather than shading areas) then the high-chroma property of viridis is actually useful. It assures that all points/lines can be seen reasonably well.
- Instead of one of the viridis family palettes one can also use a restricted range from another sequential palette. The restriction is typically necessary to assure that no colors become too light (which would make the corresponding points/lines fade into the background). The advantage is that the plot will be calmer and not so busy as with the strong hue changes of viridis. The disadvantage is that it will be harder to tell the different ends of a spectrum apart.

In short: For shading areas I would use YlGnBu. For points and lines I would go with a multi-hue sequential palette with a reasonable range of hues and relatively high chroma throughout. Maybe Mako or Rocket are a bit less busy compared to viridis.

I'm happy to play around with this a bit if you provide code for a couple of typical examples you have in mind.

**Comment by grantmcdermott** [comment]
Thanks @zeileis! Right now, we don't support heatmaps, so it's mostly for points and similar plot types. (Exception: Adding continuous color support for line plots is v. tricky with the base graphics scaffolding, so I don't plan on supporting line plots yet.)

I do like the Mako and Rocket palettes&mdash;and use them a lot in my own plots&mdash;so just from a pure aesthetic perspective it would be between those and Cividis for me. But I defer to your expert knowledge...

Let me tidy up this prototype code a little bit and push the changes to the [continuous-legend](https://github.com/grantmcdermott/tinyplot/tree/continuous-legend) branch that I'm working on. I'll ping you here when it's ready, so you can pull and play around with it on your local machine.

**Comment by zeileis** [comment]
Re: heatmaps. We don't support them \_yet\_ ;-) More seriously, for me the question is more whether we will have a conceptual difference somewhere for default continuous palette for shading areas vs. points/lines. Or whether the defaults are set per function anyway.

Re: cividis. My personal opinion is that this is a very bad choice for almost all situations. It collapses many color contrasts that would help trichromats (with "normal" color vision) to better distinguish the colors in a palette. Also, it's a weird combination between sequential and diverging palette.

Re: pinging later. :rocket:

**Comment by grantmcdermott** [comment]
Hey @zeileis I ended up doing a bit more work on the branch, so that it was ready for PR in #122. Even though I've marked it as [WIP], it's ready to kick the tires. (I just need to clean up some things internally and add tests, etc.) Feel free play around and lmk what you think.
